### PR TITLE
Break DataStore's implementation into modules

### PR DESCRIPTION
**Pull request by jgallagher**
This is a purely mechanical movement of the majority of `DataStore`'s methods into separate module files in an attempt to speed up compilation (particularly incremental compilation). On my machine, it looks like a modest success, cutting incremental compilation down by ~7 seconds for both test and non-test builds:

| Operation | main (`2a2a7f7e`) | this branch (`647e336e`) |
| --- | --- | --- |
| clean build, lib+bin | 2m 55s | 2m 48s |
| incremental build, lib+bin, `datastore.rs` modified | 45s | 38s |
| clean build, tests | 5m 5s | 4m 49s |
| incremental build, tests, `datastore.rs` modified | 1m 14s | 1m 8s |

where "lib+bin" is `cargo build -p omicron-nexus` and "tests" is `cargo build -p omicron-nexus --tests` (i.e., the build of the tests without running them). Modifying files underneath `datastore/` on this branch results in identical timings to modifying `datastore/mod.rs` (which is surprising to me, but it's quite reliable).

These timings are all debug builds; if there's interest I can collect them for release builds also.

### CODE REVIEW

**Comment by david-crespo** [comment]
Pretty cool, didn't know this would have this effect. 10% on incremental builds is pretty good!

**Comment by jgallagher** [comment]
@smklein suggested it when we were chatting about ideas as a way to speed up incremental times since diesel is one of the biggest time sinks. I don't understand why this would speed up the clean builds, though!

File: nexus/src/db/datastore/mod.rs [hunk 1]
```diff
@@ -0,0 +1,1061 @@
...
+    };
+    use omicron_test_utils::dev;
+    use std::collections::HashSet;
+    use std::net::Ipv6Addr;
+    use std::net::SocketAddrV6;
+    use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+    use std::sync::Arc;
+    use uuid::Uuid;
+
+    #[tokio::test]
```
Comment (smklein) [inline nexus/src/db/datastore/mod.rs:255]: It'll probably make sense to move some of these out too, but we can punt on that for now

**Review by smklein — APPROVED** [review]
From my own tests:

Before this patch, incremental rebuild (`touch datastore.rs && cargo build`): In the range of 19-21 seconds
 
After this patch, incremental rebuild (`touch datastore/mod.rs && cargo build`): In the range of 19-21 seconds

I'm not seeing much of a difference here - I'm on a 16 core, 32 GiB machine.

That being said, I still like this PR regardless of the performance implications, so LGTM, but it's curious that I'm not replicating your results.

**Comment by jgallagher** [comment]
Out of curiosity, what happens if you modify `datastore.rs` (e.g., add a word to the doc comment at the top of the file) instead of just `touch`'ing it?

**Pull request merged**